### PR TITLE
[PET-to-MLIR] added symbolic bounds with test case

### PR DESCRIPTION
**Pull request by Komisarczyk**

### CODE REVIEW

File: mlir/tools/mlir-pet/Lib/islNodeBuilder.cc [hunk 1]
```diff
@@ -35,13 +40,56 @@ void IslNodeBuilder::createFor(isl::ast_node forNode) {
...
   auto incrementAsInt = std::abs(createIntFromIslExpr(increment));
-  auto upperBoundAsInt = std::abs(createIntFromIslExpr(upperBound)) + 1;
 
-  auto loop =
-      MLIRBuilder_.createLoop(lowerBoundAsInt, upperBoundAsInt, incrementAsInt);
+  AffineForOp loop;
+  // symbolic upper bound
+  if (isInt(lowerBound) && isId(upperBound)) {
+    // get id from node type
+    auto upperBoundId = isl_ast_expr_get_id(upperBound.get());
```
Comment (chelini) [inline mlir/tools/mlir-pet/Lib/islNodeBuilder.cc]: can you please use C++ bindings. auto upperBoundId = isl::manage(....). this avoid isl_id_free.

File: mlir/tools/mlir-pet/Lib/islNodeBuilder.cc [hunk 2]
```diff
@@ -35,13 +40,56 @@ void IslNodeBuilder::createFor(isl::ast_node forNode) {
...
+    auto lowerBoundProcessed = std::abs(createIntFromIslExpr(lowerBound));
+    loop = MLIRBuilder_.createLoop(lowerBoundProcessed, upperBoundProcessed,
+                                   incrementAsInt);
+    isl_id_free(upperBoundId);
+  }
+
+  // both symbolic
+  else if (isId(lowerBound) && isId(upperBound)) {
+    // upperBound processing
+    auto upperBoundId = isl_ast_expr_get_id(upperBound.get());
```
Comment (chelini) [inline mlir/tools/mlir-pet/Lib/islNodeBuilder.cc]: same here

File: mlir/tools/mlir-pet/Lib/islNodeBuilder.cc [hunk 3]
```diff
@@ -35,13 +40,56 @@ void IslNodeBuilder::createFor(isl::ast_node forNode) {
...
+    auto upperBoundProcessed = isl_id_get_name(upperBoundId);
+    // lowerBound processing
+    auto lowerBoundId = isl_ast_expr_get_id(lowerBound.get());
```
Comment (chelini) [inline mlir/tools/mlir-pet/Lib/islNodeBuilder.cc]: same here

File: mlir/tools/mlir-pet/Lib/islNodeBuilder.cc [hunk 4]
```diff
@@ -35,13 +40,56 @@ void IslNodeBuilder::createFor(isl::ast_node forNode) {
...
+  // symbolic lower bound
+  else if (isId(lowerBound) && isInt(upperBound)) {
+    // get id from node type
+    auto lowerBoundId = isl_ast_expr_get_id(lowerBound.get());
+    // access string name
+    auto lowerBoundProcessed = isl_id_get_name(lowerBoundId);
+    auto upperBoundProcessed = std::abs(createIntFromIslExpr(upperBound) + 1);
+    loop = MLIRBuilder_.createLoop(lowerBoundProcessed, upperBoundProcessed,
+                                   incrementAsInt);
+    isl_id_free(lowerBoundId);
```
Comment (chelini) [inline mlir/tools/mlir-pet/Lib/islNodeBuilder.cc]: same here

File: mlir/tools/mlir-pet/Test/symBounds.mlir [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+// CHECK: #map0 = affine_map<(d0, d1) -> (d0, d1)>
+// CHECK_NEXT:#map1 = affine_map<(d0) -> (d0)>
+// CHECK_NEXT: #map2 = affine_map<() -> (0)>
+// CHECK_NEXT: #map3 = affine_map<() -> (1024)>
+// CHECK_NEXT: #map4 = affine_map<() -> (1)>
+
+// CHECK: func @scop_entry(%arg0: memref<1023x1023xf32>, %arg1:
+// memref<1023x1023xf32>, %arg2: memref<1023x1023xf32>) {
+// CHECK_NEXT:  affine.for %arag3 = 1 to 1024 {
```
Comment (chelini) [inline mlir/tools/mlir-pet/Test/symBounds.mlir]: are you sure this test passes? "arag3" looks bad

**Review by chelini — APPROVED** [review]

File: mlir/tools/mlir-pet/Lib/mlirCodegen.cc [hunk 6]
```diff
@@ -772,13 +772,83 @@ LogicalResult MLIRCodegen::verifyModule() {
...
+
+  ValueRange ubOperands = ValueRange(ub);
+  ValueRange lbOperands = {};
+  auto loop = builder_.create<AffineForOp>(builder_.getUnknownLoc(), lbOperands,
+                                           lbMap, ubOperands, ubMap, step);
+  loop.getBody()->clear();
+
+  builder_.setInsertionPointToStart(loop.getBody());
+  builder_.create<AffineTerminatorOp>(builder_.getUnknownLoc());
+  builder_.setInsertionPointToStart(loop.getBody());
```
Comment (ftynse) [inline mlir/tools/mlir-pet/Lib/mlirCodegen.cc:804]: Why is this necessary?  The builder of AffineForOp will create an AffineTerminatorOp for you, and this code erases it and creates it back again... Was this cargo-culted from the tutorial?

Comment (chelini) [reply mlir/tools/mlir-pet/Lib/mlirCodegen.cc:804]: Not necessary, thanks for pointing this out. It was taken from the tutorial. Fixed now.
